cpp/src/arrow/csv/writer.cc

```diff
@@ -462,25 +483,32 @@ class CSVWriterImpl : public ipc::RecordBatchWriter {
     return Status::OK();
   }
 
-  int64_t CalculateHeaderSize() const {
+  Result<int64_t> CalculateHeaderSize() const {
     int64_t header_length = 0;
     for (int col = 0; col < schema_->num_fields(); col++) {
       const std::string& col_name = schema_->field(col)->name();
+      int64_t escaped_count = CountQuotes(col_name);
+      if (!options_.escaping && escaped_count > 0) {
+        return Status::Invalid("Need to escape in CSV header, but escaping is disabled.");
+      }
       header_length += col_name.size();
-      header_length += CountQuotes(col_name);
+      header_length += escaped_count;
     }
     return header_length + (kQuoteDelimiterCount * schema_->num_fields());
   }
 
   Status WriteHeader() {
     // Only called once, as part of initialization
-    RETURN_NOT_OK(data_buffer_->Resize(CalculateHeaderSize(), /*shrink_to_fit=*/false));
+    int64_t header_size;
+    ASSIGN_OR_RAISE(header_size, CalculateHeaderSize());
```

Comment:
N.B. the ASSIGN_OR_RAISE macro lets you declare the variable inline (`ASSIGN_OR_RAISE(int64_t header_size, CalculateHeaderSize());`)